### PR TITLE
Implement create company and search

### PR DESCRIPTION
**Pull request by tedim52**
-Implemented __create_company - via dictionary implementation and search functions in Google class
-Added test cases for each function
-Added resource.py module to hold Resource class
-Added __hash__, __eq__, __repr__ to Googler class
-Working now on testing and implementing distance function

### CODE REVIEW

File: Google.py [hunk 1]
```diff
@@ -78,9 +80,18 @@ def __create_company(self, googlers):
...
         #TODO: Implement function to create managerial hierarchy.
-        return
+
+        # Current implementation utilizes a dictionary to represent a pseudo company hierarchy.
+        # Every Googler is a key with its value being the Googler object that represents its manager.
+        company = {}
+        for googler in googlers:
+            # For every Googler, find their manager and map googler --> manager in dict
+            # NOTE: Believe this is implicitly O(n) making this algo O(n^2) :( if you're reading this guys, im using the thinker
+            manager = next((manager for manager in googlers if manager.id == googler.manager_id), None)
```
Comment (DeanAlvarez) [inline Google.py]: This is super terse. I'll be honest I don't really know what is going on here. 

That being said I don't think the map should be <googler,googler> I think it should be <integer,googler> you could build it in o(n) just by iterating through the csv file. Alternativenly, if you want to take in a list of googlers, just map the google Id to the googler. The idea is that if the manager_id in the data maps to a user_id you can follow the chain that way. We can talk about it tomorrow in the standup.

File: Google.py [hunk 2]
```diff
@@ -18,13 +18,13 @@ class Google:
...
 
-        Calls __create_company method which implements manager hierarchy tree based on the list of
+        Calls __create_company method that implements manager hierarchy tree based on the list of
         Googlers and their attributes.
 
         Args:
             googlers: A list of Googler objects.
         """
-        self.ceo = self.__create_company(googlers)
+        self.company = self.__create_company(googlers)
```
Comment (DeanAlvarez) [inline Google.py:27]: Why the two underscores? is that a style rule thing>

File: Google.py [hunk 3]
```diff
@@ -37,8 +37,10 @@ def search(self, googler_id):
         Returns:
             googler: The Googler object attached with the googler_id.
         """
-        #TODO: Implement Googler Search function
-        return
+        googler = next((googler for googler in self.googlers if googler.id == googler_id), None)
```
Comment (DeanAlvarez) [inline Google.py]: If you implement the hashmap as user_id to googler this is o(1)

File: Googler.py [hunk 4]
```diff
@@ -37,9 +37,28 @@ def __init__(self, attributes):
...
+        return hash(self.id)
+
+    def __eq__(self, googler):
+        """Compare if two Googlers are the same person.
+
+        Two Googlers are only equal if their id's are equal to each other.
+        """
+        if isinstance(googler, type(self)):
+            return self.id == googler.id
+        return False
```
Comment (DeanAlvarez) [inline Googler.py:54]: this is good.

File: Googler.py [hunk 5]
```diff
@@ -37,9 +37,28 @@ def __init__(self, attributes):
         self.location = attributes.get("location")
         self.lowest_dir_id = attributes.get("lowest_dir_id")
         self.job_family = attributes.get("job_family")
-        self.resources = []
+        self.resources = [] #List of Resource objects from resource.py module
```
Comment (DeanAlvarez) [inline Googler.py:40]: In Kayla's dictionary file that you might use to load in it is a set of resources not a list.

File: resource.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        if not isinstance(resource, type(self)):
+            return NotImplemented
+        else:
+            return self.attr_one == other.attr_one and self.attr_two == other.attr_two
+
+    # We hash resources based off of attr_one and attr_two
+    # This technique is called cantor pairing
+    # See en.wikipedia.org/wiki/Pairing_function for more information
+    def __hash__(self):
+        return (((self.attr_one+self.attr_two+one)*(self.attr_one+self.attr_two)//two)+self.attr_two)
```
Comment (DeanAlvarez) [inline resource.py:34]: wow look at this absolutely EPIC hash function!

File: Google.py (hunk 1 above)
Comment (DeanAlvarez) [inline Google.py]:
```suggestion
            company[googler.user_id] = googler
```

**Review by DeanAlvarez — COMMENTED** [review]
We should talk about the implementation tomorrow morning

File: Google.py (hunk 2 above)
Comment (tedim52) [reply Google.py:27]: https://www.geeksforgeeks.org/private-methods-in-python/ Yeah Python has no such thing as private methods like Java but double underscore indicates it should only be used within the class.. in this case, to create the dictionary company attribute.

File: resource.py (hunk 6 above)
Comment (tedim52) [reply resource.py:34]: made it myself :)

File: Google.py (hunk 1 above)
Comment (tedim52) [reply Google.py]: I see what you're saying...  I'll make the change because then it does become O(1)

**Comment by tedim52** [comment]
-Adjusted company implementation so dictionary maps <id:Googler>, search becomes O(1)
-Implemented basic distance function off dictionary implementation with test cases

File: Google.py [hunk 7]
```diff
@@ -55,8 +55,19 @@ def distance(self, googler_one, googler_two):
...
         """
-        #TODO: Implement function to determine distance between to Googlers.
-        return
+        if(googler_one == googler_two):
+            return 0
+        googler_one_managers = set()
+        googler_two_managers = set()
+        while(len(googler_one_managers.intersection(googler_two_managers)) == 0):
+            if(googler_one.manager_id != -1):
+                googler_one_managers.add(googler_one.manager_id)
```
Comment (tedim52) [inline Google.py:64]: -1 indicates that the current googler is the "CEO" and has no manager so that the manager will only be added should they have one.

Comment (tedim52) [reply Google.py:64]: Depends on how the googler_attribute_table.csv sets up the id to manager id relationship though.

File: Google.py (hunk 2 above)
Comment (DeanAlvarez) [reply Google.py:27]: cool, thanks for sharing!

**Review by DeanAlvarez — APPROVED** [review]
Looks good, speedy change!

File: Googler.py (hunk 5 above)
Comment (tedim52) [reply Googler.py:40]: Noted. We'll talk about the file format and decide the best way to ingest the information.

File: Google_test.py [hunk 8]
```diff
@@ -3,22 +3,108 @@
...
+    def test_get_manager(self):
+        # Tests that a googles manager is retrieved
+        googler = Googler({ "id": 14, "department": 10, "cost_center":10, "manager_id":0,
+                                        "location": 10, "lowest_dir_id":10, "job_family": 10})
+        manager = self.google.company.get(googler.manager_id)
+        self.assertEqual(self.actual_manager, manager)
+
+    def test_get_ceo(self):
+        # Test that the ceo has a None manager object
+        self.assertEqual(None, None)
```
Comment (tedim52) [inline Google_test.py:50]: Going to implement this test case when we receive updated dataset and I know how user id to manager id relationship will work.

**Pull request merged**